### PR TITLE
Add cloudEvents and knative extensions

### LINKED ISSUE #3459: Extensions for camel-knative and camel-cloudevents
https://github.com/apache/camel-k-runtime/issues/752

Camel 3.15 comes with new components

- camel-knative
- camel-cloudevents

Camel K needs extensions for those, to be ported over from camel-k-runtime at
https://github.com/apache/camel-k-runtime/tree/main/camel-k-knative
https://github.com/apache/camel-k-runtime/tree/main/camel-k-cloudevents

### PR DESCRIPTION
**Pull request by zbendhiba**
Fixes #3459

### CODE REVIEW

**Comment by zbendhiba** [comment]
@lburgazzoli could you check this PR?
Is there a way to have more integration tests. I wonder if there's something we could use to simulate Knative, or do some mock.

**Comment by zbendhiba** [comment]
The vert.x native error seems to be related to this change I guess : https://github.com/vert-x3/vertx-web/commit/8dda3dec91648eddae01455b8a94cc1c0e25c864

File: integration-tests/knative-channel-consumer/pom.xml [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        <artifactId>camel-quarkus-build-parent-it</artifactId>
+        <version>2.13.0-SNAPSHOT</version>
+        <relativePath>../../poms/build-parent-it/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>camel-quarkus-integration-test-knative-channel-consumer</artifactId>
+    <name>Camel Quarkus :: Integration Tests :: Knative Consumer</name>
+    <description>Integration tests for Camel Quarkus Knative Consumer extension - Channel consumer</description>
+
+    <dependencyManagement>
```
Comment (jamesnetherton) [inline integration-tests/knative-channel-consumer/pom.xml:35]: The BOM imports can be removed.

File: integration-tests/knative-endpoint-consumer/pom.xml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        <artifactId>camel-quarkus-build-parent-it</artifactId>
+        <version>2.13.0-SNAPSHOT</version>
+        <relativePath>../../poms/build-parent-it/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>camel-quarkus-integration-test-knative-endpoint-consumer</artifactId>
+    <name>Camel Quarkus :: Integration Tests :: Knative Consumer</name>
+    <description>Integration tests for Camel Quarkus Knative Consumer extension - Endpoint consumer</description>
+
+    <dependencyManagement>
```
Comment (jamesnetherton) [inline integration-tests/knative-endpoint-consumer/pom.xml:35]: The BOM imports can be removed.

File: integration-tests/knative-event-consumer/pom.xml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        <artifactId>camel-quarkus-build-parent-it</artifactId>
+        <version>2.13.0-SNAPSHOT</version>
+        <relativePath>../../poms/build-parent-it/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>camel-quarkus-integration-test-knative-event-consumer</artifactId>
+    <name>Camel Quarkus :: Integration Tests :: Knative Consumer Event</name>
+    <description>Integration tests for Camel Quarkus Knative Consumer extension - Consuming events</description>
+
+    <dependencyManagement>
```
Comment (jamesnetherton) [inline integration-tests/knative-event-consumer/pom.xml:35]: The BOM imports can be removed.

File: integration-tests/knative-producer/pom.xml [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        <artifactId>camel-quarkus-build-parent-it</artifactId>
+        <version>2.13.0-SNAPSHOT</version>
+        <relativePath>../../poms/build-parent-it/pom.xml</relativePath>
+    </parent>
+
+    <artifactId>camel-quarkus-integration-test-knative-producer</artifactId>
+    <name>Camel Quarkus :: Integration Tests :: Knative Producer</name>
+    <description>Integration tests for Camel Quarkus Knative Producer extension</description>
+
+    <dependencyManagement>
```
Comment (jamesnetherton) [inline integration-tests/knative-producer/pom.xml:35]: The BOM imports can be removed.

File: extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    @BuildStep
+    List<UnremovableBeanBuildItem> unremovableBeans() {
+        return List.of(
+                UnremovableBeanBuildItem.beanTypes(KnativeEnvironment.class));
+    }
+
+    @BuildStep
+    List<ReflectiveClassBuildItem> reflectiveClasses() {
```
Comment (jamesnetherton) [inline extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java:57]: Maybe we should open a Camel ticket to see if we can avoid the need for reflective access in future.

Comment (lburgazzoli) [reply extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java:57]: Yeah, I think we can mark them with the `@Configurer` annotation and then we may get a reflection free bean binding

Comment (lburgazzoli) [reply extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java:57]: mh, it looks like it is the case already https://github.com/apache/camel/blob/main/components/camel-knative/camel-knative-api/src/main/java/org/apache/camel/component/knative/spi/KnativeEnvironment.java#L41

Comment (zbendhiba) [reply extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java:57]: Okay

File: integration-tests/knative-producer/pom.xml (hunk 4 above)
Comment (zbendhiba) [reply integration-tests/knative-producer/pom.xml:35]: thanks I'll do

File: extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java (hunk 5 above)
Comment (zbendhiba) [reply extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java:57]: that is not fixed yet

Comment (zbendhiba) [reply extensions/knative/deployment/src/main/java/org/apache/camel/quarkus/component/knative/deployment/KnativeProcessor.java:57]: https://issues.apache.org/jira/browse/CAMEL-18561

**Comment by zbendhiba** [comment]
@jamesnetherton @lburgazzoli  Is deleting the reflective classes mandatory for this PR?

**Comment by zbendhiba** [comment]
Let me know if something needed to explain or improve

**Comment by jamesnetherton** [comment]
> @jamesnetherton @lburgazzoli Is deleting the reflective classes mandatory for this PR?

I think it's ok to have the reflective config initially. But we should try to remove it in a future release if possible.

**Comment by zbendhiba** [comment]
> > @jamesnetherton @lburgazzoli Is deleting the reflective classes mandatory for this PR?
> 
> I think it's ok to have the reflective config initially. But we should try to remove it in a future release if possible.

I can do that for next re

**Comment by zbendhiba** [comment]
> > @jamesnetherton @lburgazzoli Is deleting the reflective classes mandatory for this PR?
> 
> I think it's ok to have the reflective config initially. But we should try to remove it in a future release if possible.

Okay I'll open an issue and add it to my tasks if this is approved for next release :)

**Review by jamesnetherton — APPROVED** [review]

**Review by lburgazzoli — APPROVED** [review]

**Pull request merged**